appside/model/dbt_worksheet_model.js

```diff
@@ -208,6 +145,78 @@ class DbtWorksheetModelFwd extends Model {
         return ret;
     }
 
+    /**
+     * Read uds from this.uds and compute the phq-9 score
+     */
+    phq_compute() {
+        // get the phq uds and tally the score
+        let phq_uds = this.uds.get_all(RESPONSE_PHQ);
+        let score = 0;
+        for(let ud of phq_uds) {
+            score += Number(ud.response);
+        }
+
+        if(score >= PHQ_LOWEST_FAIL) {
+            let i;
+            for(i=0; i<this.frames.length; i++){
+                let frame = this.frames[i];
+                // change phq result frame text
+                if(frame.template === PHQR_FRAME_TEMPLATE) {
+                    frame.instruction = PHQR_TEXT_NO;
+                    break;
+                }
+            }
+            // delete study and app frames, leaving only the end frame
+            let start_deleting_idx = i + 3;
+            for(; i<this.frames.length; i++) {
+                let frame = this.frames[i];
+                if(frame.template === END_FRAME_TEMPLATE) {
+                    frame.set_passed_phq(false);
+                    break;
+                }
+            }
+            let num_to_delete = i - start_deleting_idx;
+            this.frames.splice(start_deleting_idx, num_to_delete);
+        }
+    }
+
+    /**
+     * Build phq frames
+     * @return the question frame and the results frame
+     */
+    build_phq_frames() {
+        let ret = [];
+
+        // the frame with the questions
+        let q_frame = {};
+        q_frame.template = PHQ_FRAME_TEMPLATE;
+        q_frame.title = PHQ_TITLE;
+        q_frame.instruction = PHQ_TEXT;
+        q_frame.questions = PHQ_QUESTIONS;
+        q_frame.response_name = RESPONSE_PHQ;
+        ret.push(new FormFrame(q_frame, this.logger));
+
+        for(let question of q_frame.questions) {
+            let text = question[0];
+            let ud = new UserData(text, '', [], q_frame.response_name);
+            this.uds.add(ud);
+        }
+        ret.push(new BlockerFrame());
+
+        // the frame with the results
+        let r_frame = {};
+        r_frame.template = PHQR_FRAME_TEMPLATE;
+        r_frame.title = PHQR_TITLE;
+        // this field will be changed after the fact if participant fails the screen
```

Comment:
rather than changing the field directly, ideally we'd change a flag on the frame and tell it to change its own field. But this is form frame which is too generalized to have a "passed scren" flag. So let this stand.